pkg/tempdir/tempdir.go

```diff
@@ -0,0 +1,306 @@
+
+	rootDir/
+	    .lock-temp-dir-ABC           (instance lock for temp-dir-ABC)
+	    temp-dir-ABC/
+	        file1
+	        file3
+	    .lock-temp-dir-XYZ           (instance lock for temp-dir-XYZ)
+	    temp-dir-XYZ/
+	        file2
+*/
+const (
+	// TempDirPrefix is the prefix used for creating temporary directories.
+	TempDirPrefix = "temp-dir-"
+	// TempdirLockPrefix is the prefix used for creating lock files for temporary directories.
+	TempdirLockPrefix = ".lock-"
+)
+
+// TempDir represents a temporary directory that is created in a specified root directory.
+// It manages the lifecycle of the temporary directory, including creation, locking, and cleanup.
+// Each TempDir instance is associated with a unique subdirectory in the root directory.
+type TempDir struct {
+	rootDir string
+
+	tempDirPath string
+	// tempDirLock is a lock file (e.g., rootDir/.lock-temp-dir-XYZ) specific to this
+	// TempDir instance, indicating it's in active use.
+	tempDirLock     *lockfile.LockFile
+	tempDirLockPath string
+}
+
+type staleDir struct {
+	lockPath    string
+	tempDirPath string
+}
+
+// listPotentialStaleDirs scans the rootDir for directories that might be stale temporary directories.
+// It identifies directories with the TempDirPrefix and their corresponding lock files with the TempdirLockPrefix.
+// If a directory does not have a corresponding lock file, it is considered stale.
+// The function returns a list of staleDir objects, each containing the paths to the directory and its lock file.
+func (td *TempDir) listPotentialStaleDirs() ([]staleDir, error) {
+	dirs := make(map[string]struct{})
+	locks := make(map[string]struct{})
+
+	dirContent, err := os.ReadDir(td.rootDir)
+	if err != nil {
+		return nil, fmt.Errorf("error reading temp dir %s: %w", td.rootDir, err)
+	}
+
+	for _, entry := range dirContent {
+		path := filepath.Join(td.rootDir, entry.Name())
+		if strings.HasPrefix(entry.Name(), TempDirPrefix) {
+			dirs[path] = struct{}{}
+			continue
+		}
+		if strings.HasPrefix(entry.Name(), TempdirLockPrefix) {
+			locks[path] = struct{}{}
+		}
+	}
+
+	out := []staleDir{}
+
+	for path := range dirs {
+		base := filepath.Base(path)
+		lockPath := filepath.Join(td.rootDir, TempdirLockPrefix+base)
+		if _, ok := locks[lockPath]; ok {
+			delete(locks, lockPath)
+		} else {
+			lockPath = ""
+		}
+		out = append(out, staleDir{
+			tempDirPath: path,
+			lockPath:    lockPath,
+		})
+	}
+
+	for path := range locks {
+		out = append(out, staleDir{
+			tempDirPath: "",
+			lockPath:    path,
+		})
+	}
+
+	return out, err
+}
+
+// RecoverStaleDirs identifies and removes stale temporary directories in the root directory.
+// A directory is considered stale if its lock file can be acquired (indicating no active use).
+// The function attempts to remove both the directory and its lock file.
+// Any errors encountered during the recovery process are collected and returned as a single error.
+func (td *TempDir) RecoverStaleDirs() error {
```

Comment:
This is not related to the mutable (“instance”) fields of the object.

I’d truly find it much cleaner to have a separate handling of the “temp dir root”, as one object, and “instance”, as another. (Maybe we don’t need a root _object_, and just pass a root path. Anyway.) The two have different lifetimes, and different concurrency regimes.